### DISCUSSION TITLE
Is it possible to clear a specific template cache using a specific key since the template cache refactor in Craft CMS 3.5+?

### DISCUSSION THREAD

**Discussion by jamesmacwhite** [craftcms/cms#15745] [Q&A] [answered]
I have a scenario where we I'm using the `{% cache %}` tag for the response a web request which is being refreshed by a queue job every 15 minutes. The template is currently being cached by a fixed time value. However, rather than set a expiry on the template, I'd rather have the template cache use the default duration and specify no specific parameters for expiring and be deleted upon the queue job refresh, given I'm setting a specific key and this queue job trigger the web request offloaded from the front end. However, is there any easy way to delete a specific template cache key since the refactoring of template caching?

In Craft CMS 3, you could do:

```php
Craft::$app->getTemplateCaches()->deleteCachesByKey('key');
```

This however was deprecated and doesn't exist in Craft CMS 5, is there a equivalent way to achieve this?

**Comment by brandonkelly** [comment, marked as answer]
You can register a custom cache invalidation tag to the cache:

```twig
{% cache %}
  {% do craft.app.elements.collectCacheTags(['custom-tag']) %}
  ...
{% endcache %}
```
Then clear caches by that tag in your queue job:

```php
use yii\caching\TagDependency;

TagDependency::invalidate(Craft::$app->cache, 'custom-tag');
```

For extra credit, you can register the tag with the Caches utility, which will add a checkbox for the tag under “Invalidate Data Caches”:

```php
use craft\base\Event;
use craft\events\RegisterCacheOptionsEvent;
use craft\utilities\ClearCaches;

Event::on(ClearCaches::class, ClearCaches::EVENT_REGISTER_TAG_OPTIONS, function(RegisterCacheOptionsEvent $event) {
    $event->options[] = [
        'tag' => 'custom-tag',
        'label' => 'My Custom Tag',
    ];
});
```

That will also add an `invalidate-tags/custom-tag` CLI command, which you could call via a cron job rather than going the queue job route, if you want.

**Reply by jamesmacwhite** [reply]
@brandonkelly Thanks for this! This is great additional detail and I'll look at implementing custom tags for specific template invalidation for this use case!
